### PR TITLE
Improve error reporting when custom error handler is used

### PR DESCRIPTION
**Pull request by SimonFrings**
See https://github.com/reactphp/socket/pull/290
Builds on top of 20bd61a and #37.

### CODE REVIEW

File: src/functions.php [hunk 1]
```diff
@@ -104,13 +104,21 @@
...
+        // Match errstr from PHP's warning message.
+        // stream_filter_append() expects parameter 1 to be resource,...
+        $errstr = $error;
+    });
+
+    $ret = \stream_filter_append($stream, register(), $read_write, $callback);
+
+    \restore_error_handler();
 
     // PHP 8 throws above on type errors, older PHP and memory issues can throw here
```
Comment (clue) [inline src/functions.php:124]: This suggest PHP 8 type errors aren't currently taken care of as the `stream_filter_append()` would throw, can you look into this?

File: src/functions.php [hunk 2]
```diff
@@ -147,13 +155,21 @@ function append($stream, $callback, $read_write = STREAM_FILTER_ALL)
...
+        // Match errstr from PHP's warning message.
+        // stream_filter_prepend() expects parameter 1 to be resource,...
+        $errstr = $error;
+    });
+
+    $ret = \stream_filter_prepend($stream, register(), $read_write, $callback);
+
+    \restore_error_handler();
 
     // PHP 8 throws above on type errors, older PHP and memory issues can throw here
```
Comment (clue) [inline src/functions.php:181]: This suggest PHP 8 type errors aren't currently taken care of as the `stream_filter_prepend()` would throw, can you look into this?

File: src/functions.php [hunk 3]
```diff
@@ -301,10 +326,20 @@ function fun($filter, $parameters = null)
...
+        // stream_filter_remove() expects parameter 1 to be resource,...
+        $errstr = $error;
+    });
+
+    $ret = \stream_filter_remove($filter);
+
+    \restore_error_handler();
+
+    if ($ret === false) {
         // PHP 8 throws above on type errors, older PHP and memory issues can throw here
```
Comment (clue) [inline src/functions.php:359]: This suggest PHP 8 type errors aren't currently taken care of as the `stream_filter_remove()` would throw, can you look into this?

File: tests/FilterTest.php [hunk 4]
```diff
@@ -335,37 +335,66 @@ public function testAppendThrowsShouldTriggerEndButIgnoreExceptionDuringEnd()
...
+        set_error_handler(function ($_, $errstr) use (&$error) {
+            $error = $errstr;
+        });
+
         StreamFilter\remove(false);
+
+        $error = null;
+        set_error_handler(function ($_, $errstr) use (&$error) {
+            $error = $errstr;
+        });
```
Comment (clue) [inline tests/FilterTest.php]: Copy pasta?

File: tests/FilterTest.php [hunk 5]
```diff
@@ -335,37 +335,66 @@ public function testAppendThrowsShouldTriggerEndButIgnoreExceptionDuringEnd()
...
     }
 
     /**
      * @depends testAppendEndEventWillBeCalledOnRemove
      */
-    public function testRemoveThrowsWhenAppendThrowsOnEvent()
+    public function testRemoveThrowsWhenAppendThrowsOnEventWithoutCallingCustomErrorHandler()
```
Comment (clue) [inline tests/FilterTest.php]: Copy pasta?

**Review by clue — CHANGES_REQUESTED** [review]
@SimonFrings Thanks for looking into this! :+1:

File: tests/FilterTest.php (hunk 5 above)
Comment (SimonFrings) [reply tests/FilterTest.php]: Don't know what I was thinking here but this is clearly wrong.

File: tests/FilterTest.php (hunk 4 above)
Comment (SimonFrings) [reply tests/FilterTest.php]: Small brain moment, fixed it.

File: src/functions.php (hunk 1 above)
Comment (SimonFrings) [reply src/functions.php:124]: Done 👍

File: src/functions.php (hunk 2 above)
Comment (SimonFrings) [reply src/functions.php:181]: Done 👍

File: src/functions.php (hunk 3 above)
Comment (SimonFrings) [reply src/functions.php:359]: Done 👍

**Review by clue — APPROVED** [review]
@SimonFrings Thanks for the update, changes LGTM! :shipit:

**Pull request merged**